File: python/src/pay_kit/protocols/x402/exact/extensions.py
"""x402 v2 ``extensions`` echo-and-append support.

Mirrors the Rust spine ``PaymentExtensions`` and friends in
``rust/crates/x402/src/protocol/schemes/exact/types.rs`` line-for-line:

* The ``extensions`` object rides on BOTH wire envelopes. The inbound
  ``PAYMENT-REQUIRED`` challenge advertises it (untyped passthrough); the
  outbound ``PAYMENT-SIGNATURE`` credential echoes it back with any required
  client-side fields filled in (e.g. ``payment-identifier.info.id``).
* The client MUST echo at least the info it received, MAY append additional
  info, and MUST NOT delete or overwrite existing info (x402 v2 §5.1.2). Unknown
  extensions are re-emitted verbatim so forward-compatible payloads survive.
* An absent extensions object is OMITTED on the wire, never serialized as
  ``null`` or an empty ``{}`` (rust ``skip_serializing_if = "Option::is_none"``
  plus :func:`extensions_is_empty`).

The spec JSON key for the payment-identifier extension is kebab-case
``payment-identifier`` (rust ``#[serde(rename = "payment-identifier")]``); its
``info`` block is camelCase ``{ required?, id? }`` and the server-published
``schema`` is echoed verbatim.
"""

from __future__ import annotations

import copy
import re
import secrets
from typing import Any, cast

__all__ = [
    "PAYMENT_IDENTIFIER_KEY",
    "PAYMENT_IDENTIFIER_ID_PATTERN",
    "PaymentIdentifierError",
    "generate_payment_identifier_id",
    "echo_extensions",
    "requires_payment_identifier",
    "with_payment_identifier_id",
    "extensions_is_empty",
    "extract_payment_identifier_id",
    "verify_payment_identifier",
]


class PaymentIdentifierError(ValueError):
    """Raised when a route requires a payment-identifier the credential lacks.

    The coinbase x402 v2 payment_identifier spec maps this to HTTP 400. The
    message carries the canonical ``payment-identifier`` token so the
    conformance reject classifier maps it to ``payment-identifier-required``.
    """


#: Wire key for the payment-identifier extension (rust
#: ``#[serde(rename = "payment-identifier")]``).
PAYMENT_IDENTIFIER_KEY = "payment-identifier"

#: Spec pattern for a payment-identifier ``info.id`` (coinbase x402 v2
#: payment_identifier.md). The canonical Solana shape ``pay_`` + 32 hex
#: (36 chars total) satisfies this. Compiled once and reused by the server
#: reject gate.
PAYMENT_IDENTIFIER_ID_PATTERN = re.compile(r"^[A-Za-z0-9_-]{16,128}$")


def generate_payment_identifier_id() -> str:
    """Generate a fresh ``pay_``-prefixed idempotency id.

    ``pay_`` + 32 lowercase hex chars (36 total), satisfying the spec pattern
    ``^[A-Za-z0-9_-]{16,128}$`` and the canonical Solana
    ``^pay_[a-zA-Z0-9_-]{10,120}$`` shape. Mirrors rust
    ``generate_payment_identifier_id`` (16 CSPRNG bytes hex-encoded).

    Per the spec, callers MUST reuse the same id across retries of the same
    logical request so the server can return a cached 200 instead of charging
    twice.
    """
    return "pay_" + secrets.token_bytes(16).hex()


def echo_extensions(
    inbound: dict[str, Any] | None,
) -> dict[str, Any] | None:
    """Deep-copy the server's advertised extensions for the outbound credential.

    Mirrors rust ``PaymentExtensions::echoing``: returns ``None`` when the
    server advertised nothing (so the build omits the ``extensions`` object
    entirely), else a deep copy so unknown keys round-trip verbatim and the
    caller can mutate the copy without touching the inbound challenge.
    """
    if inbound is None:
        return None
    return cast("dict[str, Any]", copy.deepcopy(inbound))


def requires_payment_identifier(extensions: dict[str, Any] | None) -> bool:
    """``payment-identifier.info.required == True``.

    Mirrors rust ``PaymentExtensions::requires_payment_identifier``.
    """
    if not isinstance(extensions, dict):
        return False
    pid = extensions.get(PAYMENT_IDENTIFIER_KEY)
    if not isinstance(pid, dict):
        return False
    info = pid.get("info")
    if not isinstance(info, dict):
        return False
    return info.get("required") is True


def with_payment_identifier_id(
    extensions: dict[str, Any] | None,
    payment_id: str,
) -> dict[str, Any]:
    """Set the client-side ``payment-identifier.info.id`` without overwriting.

    Mirrors rust ``PaymentExtensions::with_payment_identifier_id``: creates the
    payment-identifier entry if the server did not advertise one (uncommon but
    spec-allowed), preserves the server's ``info.required`` and published
    ``schema`` verbatim, and only appends the client ``id``. Operates on a deep
    copy so the inbound challenge is never mutated.
    """
    next_ext: dict[str, Any] = cast("dict[str, Any]", copy.deepcopy(extensions)) if isinstance(extensions, dict) else {}
    pid = next_ext.get(PAYMENT_IDENTIFIER_KEY)
    if not isinstance(pid, dict):
        pid = {}
    info = pid.get("info")
    if not isinstance(info, dict):
        info = {}
    info["id"] = payment_id
    pid["info"] = info
    next_ext[PAYMENT_IDENTIFIER_KEY] = pid
    return next_ext


def extensions_is_empty(extensions: dict[str, Any] | None) -> bool:
    """True when the extensions object carries no keys.

    Mirrors rust ``PaymentExtensions::is_empty``. Callers use this to avoid
    emitting an empty ``extensions: {}`` on the outbound envelope.
    """
    if not isinstance(extensions, dict):
        return True
    return len(extensions) == 0


def extract_payment_identifier_id(extensions: dict[str, Any] | None) -> str | None:
    """Read ``payment-identifier.info.id`` off a credential's extensions, if any."""
    if not isinstance(extensions, dict):
        return None
    pid = extensions.get(PAYMENT_IDENTIFIER_KEY)
    if not isinstance(pid, dict):
        return None
    info = pid.get("info")
    if not isinstance(info, dict):
        return None
    value = info.get("id")
    return value if isinstance(value, str) else None


def verify_payment_identifier(
    extensions: dict[str, Any] | None,
    *,
    required: bool,
) -> None:
    """Server reject gate: enforce a required, valid payment-identifier id.

    When the route advertised ``payment-identifier`` with ``info.required = true``
    (``required=True`` here), the echoed credential MUST carry a valid
    ``pay_``-shaped ``info.id`` matching ``^[A-Za-z0-9_-]{16,128}$``. A missing,
    empty, or pattern-violating id raises :class:`PaymentIdentifierError`
    (coinbase x402 v2 spec: HTTP 400). A no-op when ``required`` is false.

    Mirrors the rust spine ``PaymentExtensions::requires_payment_identifier`` +
    the server reject layered on the accepted-vs-route checks, and the TS
    reference reject gate in ``harness/src/conformance/x402.ts``.
    """
    if not required:
        return
    payment_id = extract_payment_identifier_id(extensions)
    if payment_id is None or payment_id == "":
        raise PaymentIdentifierError("payment-identifier required but credential echoed no id")
    if not PAYMENT_IDENTIFIER_ID_PATTERN.match(payment_id):
        raise PaymentIdentifierError(
            f"payment-identifier id is invalid: {payment_id} does not match ^[A-Za-z0-9_-]{{16,128}}$"
        )

File: go/protocols/x402/extensions.go
package x402

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"regexp"
	"sort"
)

// x402 v2 extensions (mirrors rust
// rust/crates/x402/src/protocol/schemes/exact/types.rs PaymentExtensions
// et al.). The `extensions` object rides on BOTH the inbound
// PAYMENT-REQUIRED challenge and the outbound PAYMENT-SIGNATURE
// credential. The client echoes the inbound object back, fills required
// client-side fields (payment-identifier.info.id), preserves unknown
// extensions verbatim (forward-compat echo-and-append, coinbase x402 v2
// §5.1.2), and omits the object entirely when the server advertised none.

// PaymentIdentifierKey is the spec JSON key for the payment-identifier
// extension (rust #[serde(rename = "payment-identifier")]).
const PaymentIdentifierKey = "payment-identifier"

// paymentIdentifierIDPattern is the spec id constraint
// ^[A-Za-z0-9_-]{16,128}$ (rust types.rs:488).
var paymentIdentifierIDPattern = regexp.MustCompile(`^[A-Za-z0-9_-]{16,128}$`)

// PaymentIdentifierInfo carries the client/server fields of the
// payment-identifier extension. camelCase wire (rust types.rs:483-493).
type PaymentIdentifierInfo struct {
	// Required is the server-side flag: when true and Id is missing the
	// server returns 400 (rust types.rs:484-487).
	Required *bool `json:"required,omitempty"`
	// Id is the client-side idempotency key. Must match
	// ^[A-Za-z0-9_-]{16,128}$. Canonical Solana uses a pay_ prefix.
	Id string `json:"id,omitempty"`
}

// PaymentIdentifierExtension is the payment-identifier extension. The
// client echoes it into the outbound PAYMENT-SIGNATURE with info.id
// populated (rust types.rs:500-507).
type PaymentIdentifierExtension struct {
	Info PaymentIdentifierInfo `json:"info"`
	// Schema is the JSON Schema published by the server. Echoed verbatim
	// per x402 v2 §5.1.2 (rust types.rs:505-506).
	Schema json.RawMessage `json:"schema,omitempty"`
}

// PaymentExtensions is the typed view over the v2 `extensions` object.
// The payment-identifier extension is fielded out directly; unknown
// extensions flow through Other verbatim so the echo-and-append rule
// (§5.1.2) does not drop forward-compatible payloads. Mirrors rust
// PaymentExtensions { payment_identifier, #[serde(flatten)] other }
// (types.rs:513-528).
type PaymentExtensions struct {
	// PaymentIdentifier is the payment-identifier idempotency extension.
	PaymentIdentifier *PaymentIdentifierExtension
	// Other holds extensions this SDK does not type natively, captured
	// during echo and re-emitted verbatim (rust #[serde(flatten)] other).
	Other map[string]json.RawMessage
}

// MarshalJSON renders the payment-identifier extension under its kebab-case
// key alongside any verbatim unknown extensions, mirroring rust's rename +
// flatten. Go has no native serde flatten, so the two are merged by hand.
func (p PaymentExtensions) MarshalJSON() ([]byte, error) {
	out := make(map[string]json.RawMessage, len(p.Other)+1)
	for k, v := range p.Other {
		out[k] = v
	}
	if p.PaymentIdentifier != nil {
		raw, err := json.Marshal(p.PaymentIdentifier)
		if err != nil {
			return nil, err
		}
		out[PaymentIdentifierKey] = raw
	}
	return json.Marshal(out)
}

// UnmarshalJSON splits the payment-identifier extension out of the object
// and captures every other key verbatim into Other, mirroring rust's
// rename + flatten (echo-and-append: unknown extensions survive byte-equal).
func (p *PaymentExtensions) UnmarshalJSON(data []byte) error {
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return err
	}
	p.PaymentIdentifier = nil
	p.Other = nil
	for k, v := range raw {
		if k == PaymentIdentifierKey {
			var ext PaymentIdentifierExtension
			if err := json.Unmarshal(v, &ext); err != nil {
				return err
			}
			p.PaymentIdentifier = &ext
			continue
		}
		if p.Other == nil {
			p.Other = make(map[string]json.RawMessage)
		}
		p.Other[k] = append(json.RawMessage(nil), v...)
	}
	return nil
}

// IsEmpty reports whether no fields are populated. Callers use it to avoid
// emitting an empty `extensions: {}` object on outbound envelopes (rust
// PaymentExtensions::is_empty, types.rs:533-535).
func (p *PaymentExtensions) IsEmpty() bool {
	if p == nil {
		return true
	}
	return p.PaymentIdentifier == nil && len(p.Other) == 0
}

// RequiresPaymentIdentifier reports whether
// payment-identifier.info.required == true (rust
// requires_payment_identifier, types.rs:538-543).
func (p *PaymentExtensions) RequiresPaymentIdentifier() bool {
	if p == nil || p.PaymentIdentifier == nil {
		return false
	}
	return p.PaymentIdentifier.Info.Required != nil && *p.PaymentIdentifier.Info.Required
}

// PaymentIdentifierID returns the echoed client-side id, or "" when absent.
func (p *PaymentExtensions) PaymentIdentifierID() string {
	if p == nil || p.PaymentIdentifier == nil {
		return ""
	}
	return p.PaymentIdentifier.Info.Id
}

// WithPaymentIdentifierID sets (or overwrites) the client-side
// payment-identifier.info.id, creating the extension entry if the server
// did not advertise one, preserving server-side info (required) and schema
// verbatim (rust with_payment_identifier_id, types.rs:548-553).
func (p *PaymentExtensions) WithPaymentIdentifierID(id string) {
	if p.PaymentIdentifier == nil {
		p.PaymentIdentifier = &PaymentIdentifierExtension{}
	}
	p.PaymentIdentifier.Info.Id = id
}

// Keys returns the sorted top-level extension keys present on the object
// (the payment-identifier key plus any verbatim unknown keys). Used by the
// conformance oracle to assert the echo-and-append key set.
func (p *PaymentExtensions) Keys() []string {
	if p == nil {
		return nil
	}
	keys := make([]string, 0, len(p.Other)+1)
	if p.PaymentIdentifier != nil {
		keys = append(keys, PaymentIdentifierKey)
	}
	for k := range p.Other {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

// EchoExtensions deep-copies the inbound extensions blob so unknown keys
// round-trip verbatim, mirroring rust PaymentExtensions::echoing
// (types.rs:559-565). Returns nil when the server advertised no extensions
// (so the outbound omits the object). The inbound is provided as the raw
// JSON object carried on the PAYMENT-REQUIRED challenge.
func EchoExtensions(inbound json.RawMessage) (*PaymentExtensions, error) {
	if len(inbound) == 0 || string(inbound) == "null" {
		return nil, nil
	}
	var ext PaymentExtensions
	if err := json.Unmarshal(inbound, &ext); err != nil {
		return nil, err
	}
	return &ext, nil
}

// IsValidPaymentIdentifierID reports whether id matches the spec pattern
// ^[A-Za-z0-9_-]{16,128}$ (rust types.rs:488).
func IsValidPaymentIdentifierID(id string) bool {
	return paymentIdentifierIDPattern.MatchString(id)
}

// GeneratePaymentIdentifierID generates a fresh pay_-prefixed idempotency
// id (32 lowercase hex chars after the prefix; 36 total), satisfying the
// spec pattern ^[A-Za-z0-9_-]{16,128}$ and the canonical Solana
// ^pay_[a-zA-Z0-9_-]{10,120}$ shape (rust generate_payment_identifier_id,
// types.rs:575-585). Callers MUST reuse the same id across retries of the
// same logical request so the server can return a cached 200 instead of
// charging twice.
func GeneratePaymentIdentifierID() string {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		panic("x402: getrandom CSPRNG failure: " + err.Error())
	}
	return "pay_" + hex.EncodeToString(b[:])
}
